<pre><b>
Protocol - set of rules to communicate between machines.
Example - Grammar is a protocol for a language. You have to follow grammar while speaking so that others can understand your communication.
USB port - follows usb protocol so that when we plug usb, it reads from usb.

Simple Mail Transfer Protocol - to communicate with a mail server, we have to follow SMTP. This handles sending and trasferring mail. Most of the mail servers support SMTP. Different servers uses different port numbers.

SMTP is a part of python network library. It is a part of Application Layer.

SSL/TLS - standard encryption presents in TCP(Transmission Control Protocol) communication.

SSL - Secure Socket Layer
TLS - Transport Layer Security

Steps to communicate using SMTP:
    1. Create SMTP client
    2. We have to have host(domain name), port number
    
We have to pass:    
    Host - SMTP server, IP address/Domain(gmail.com)
    Port - SMTP server will be listening(587 for tls, 465 for ssl with gmail)
    Sender - String of sender email id
    Password - Password string for sender email
    Receiver - String or list of string of to email id(s)
    Message - Actual message that we want to pass(It is also a string)
    Subject - optional

SMTP communication is not secure. So we have to find a way to send the mail in unsecure way. Just check with the email server for making it less secure.

For network call, always have a try/except block
</pre></b>

In [3]:
import smtplib

In [4]:
email_id = "xyz@gmail.com"
pwd = "dummy-pwd"
to = "sender email"
message = "test subject"

# Creating the server with host and port
server = smtplib.SMTP('smtp.gmail.com',587)
server.starttls() # Secure the communication

# For SSL, comment the above 2 lines and uncomment the below line
# server = smtplib.SMTP_SSL('smtp.gmail.com',465)
server.login(email_id,pwd)
server.sendmail(email_id,to,message)
server.quit()

SMTPAuthenticationError: (535, b'5.7.8 Username and Password not accepted. Learn more at\n5.7.8  https://support.google.com/mail/?p=BadCredentials z126sm15469026pfc.94 - gsmtp')

<pre>
By default, you will get the error
<b>SMTPAuthenticationError: (535, b'5.7.8 Username and Password not accepted. Learn more at\n5.7.8  https://support.google.com/mail/?p=BadCredentials z126sm15469026pfc.94 - gsmtp')</b>

Google: less secure apps gmail(https://myaccount.google.com/lesssecureapps)
switch on this. We have to make the gmail less secure in order to send mail through smtp.
We have to turn off the 2 factor authentication also
</pre>

In [8]:
email_id = "test@gmail.com"
pwd = "pwd"
to = "to@gmail.com"
message = "test subject"

# Creating the server with host and port
server = smtplib.SMTP('smtp.gmail.com',587)
server.starttls() # Secure the communication

# For SSL, comment the above 2 lines and uncomment the below line
# server = smtplib.SMTP_SSL('smtp.gmail.com',465)
server.login(email_id,pwd)
server.sendmail(email_id,to,message)
server.quit()

(221, b'2.0.0 closing connection 5sm17745187pfw.25 - gsmtp')

<pre><b>
To send mails to multiple accounts
</b></pre>

In [12]:
email_id = "from@gmail.com"
pwd = "pwd"
emails = ["user1@gmail.com","user2@gmail.com"]
message = "new mail"

# Creating the server with host and port
server = smtplib.SMTP('smtp.gmail.com',587)
server.starttls() # Secure the communication

# For SSL, comment the above 2 lines and uncomment the below line
# server = smtplib.SMTP_SSL('smtp.gmail.com',465)
server.login(email_id,pwd)
server.sendmail(email_id,emails,message)
server.quit()

(221, b'2.0.0 closing connection x18sm16984114pfc.93 - gsmtp')

### SSL

In [6]:
email_id = "from@gmail.com"
pwd = "pwd"
to = "to@gmail.com"
message = "test subject"

# Creating the server with host and port
server = smtplib.SMTP_SSL('smtp.gmail.com',465)

# For SSL, comment the above 2 lines and uncomment the below line
# server = smtplib.SMTP_SSL('smtp.gmail.com',465)
server.login(email_id,pwd)
server.sendmail(email_id,to,message)
server.quit()

(221, b'2.0.0 closing connection f43sm15538210pjg.35 - gsmtp')

<pre><b>
Mail with subject
</b></pre>

In [13]:
from_email_id = "from@gmail.com"
pwd = "pwd"
to = "to@gmail.com"
subject = "test subject"
message = "Hello from Jupyter"

# Creating the server with host and port
server = smtplib.SMTP_SSL('smtp.gmail.com',465)
server.login(email_id,pwd)
# \r - carriage return - previous data gets appended
"""
body = "\r\n".join(['To:{}'.format(to), 'From:{}'.format(from_email_id),
                    'Subject:{}'.format(subject), 'Message: {}'.format(message)])
"""
body = "\r\n".join(['To:%s'%to, 'From:%s'%from_email_id,
                    'Subject:%s'%subject,message])

try:
    server.sendmail(from_email_id,to,body)
    print("Mail sent")
except Exception as e:
    print(str(e))
server.quit()

Mail sent


(221, b'2.0.0 closing connection x12sm16810281pff.48 - gsmtp')

<pre>
b represents the data returned as a byte
(221, b'2.0.0 closing connection x12sm16810281pff.48 - gsmtp')
</pre>